carbon-design-system/carbon-addons-iot-react
# [DashboardEditor] base functionality

Prereq PRs: #1692 #1688 #1686 
The diff will get better once the above PRs are merged

Part of #1372 #1560 #1561 #1562 
- There will be more iterations to add the custom options for each card type
- internal design: https://ibm.invisionapp.com/share/KWO0FZ5BHQ9#/screens/319695912

**Summary**

Includes: 
- integrates CardCodeEditor for individual card JSON configuration
  - Basic JSON validation (empty string, invalid JSON)
- Basic form options that update the card JSON: Title, Description, Size
  - Size includes the currently defined allowed card sizes per card type
- Tabs added for future use of 'Basics' and 'Advanced' card config form items
- PageTitleBar added as the header with callbacks and i18n attached for customization
- `notification` prop to allow for custom notifications such as error, info, success
- DashboardGrid added to show the dashboard preview
- Default cards are now shown in the DashboardGrid when added with basic empty data states
- New card actions:
  - 'Clone card' will make a copy of the card with a new unique id
  - 'Delete card' will remove the card
- `renderHeader` prop allows for a custom header
- `renderCardPreview` prop allows for custom rendering of cards

**Change List (commits, features, bugs, etc)**

- Exports DashboardEditor
- Added unit, function, integration tests, and stories

**Acceptance Test (how to verify the PR)**

Quite a few things to test here
1. Go to default story and add multiple cards of various types. Ensure they all look and feel correct. 
2. Attempt to drag the cards to make sure DashboardGrid is integrated correctly. 
3. After adding a new card, you should see the card edit form as that card is now 'selected'.
4. Edit all 3 fields and ensure you see the changes occurring in the card preview
5. Open the JSON editor by clicking the button at the bottom of the card edit form
6. Edit the JSON and ensure that the changes are reflected in the card, either visually or by looking at the React component debugger
7. While in the editor, try to add invalid JSON such as empty string, `1234`, etc and ensure that the editor won't let you save invalid JSON
8. Click on the card's overflow menu and attempt each action to ensure they work correctly (clone card should add a duplicate card, delete card should remove the card)
9. Check the customHeader and customCardRenderer stories and attempt some variations with it to ensure they are rendering the custom content correctly
10. Go to the notifications story and try passing different kinds of nodes. It currently allows any nodes, but InlineNotifications make the most sense there
11. Clicking a card should select the card
